### PR TITLE
new commits (#6)

### PR DESCRIPTION
**Pull request by 0b01**
* refactor: State => ThreadState, misc Cow

* fix:tests

* refactor: move threadstate string to cow

* refactor: misc

* fix: js client impl

* fix: js impl

* Fix small buffer size

* feature: add pretty graphics

* fix: share store across threads

### CODE REVIEW

File: src/bin/server/server.rs [hunk 1]
```diff
@@ -72,7 +75,12 @@ pub fn run_server(host: &str, port: &str, settings: &Settings) {
         let (rdr, wtr) = socket.split();
         let lines = lines(BufReader::new(rdr));
         let responses = lines.map(move |line| {
-            let resp = handler::gen_response(line.clone(), &mut state.borrow_mut());
+            #[cfg(feature="debug")]
```
Comment (Ameobea) [inline src/bin/server/server.rs]: How does this feature work?  Is this meant to be for debugging during development rather than at runtime?

**Review by Ameobea — APPROVED** [review]
LGTM :+1:

File: src/bin/server/server.rs (hunk 1 above)
Comment (0b01) [reply src/bin/server/server.rs]: During development. It's a compiler flag. https://doc.rust-lang.org/book/first-edition/conditional-compilation.html

Comment (Ameobea) [reply src/bin/server/server.rs]: Yeah, I'm familiar with conditional compilation, I just wanted to know if it's something that we wanted to worry about in deployments for tracking down problems at runtime or it it was only useful during development.

**Comment by Ameobea** [comment]
@rickyhan do you want me to handle this merge conflict?

File: src/bin/server/server.rs (hunk 1 above)
Comment (0b01) [reply src/bin/server/server.rs]: Probably a good idea to remove it TBH

Comment (Ameobea) [reply src/bin/server/server.rs]: If it's useful during dev, I'm fine with keeping it in.  I was just trying to understand the reasoning behind adding it.

Comment (0b01) [reply src/bin/server/server.rs]: The idea is to toggle a flag and have all sort of useful printout to debug with. Though chances are the next bug will be a lot more non-obvious so it's really unnecessary to interleave code with those debug blocks.

**Pull request merged**